### ISSUE TITLE
CRYPTOPP_DISABLE_ASM is broken in 8.9.0 for MSVC x64 builds

### ISSUE THREAD

**Issue by manxorist** [weidai11/cryptopp#1240]
`#define CRYPTOPP_DISABLE_ASM 1` does not disable all assembler / builtins / intrinsics (as documented here: https://github.com/weidai11/cryptopp/blob/CRYPTOPP_8_9_0/config_asm.h#L30).

`cpu.cpp` on x86-64 Windows MSVC builds still wants `CPUID64` and `XGETBV64` which are only implemented in `x64dll.asm` (see https://github.com/weidai11/cryptopp/blob/843d74c7c97f9e19a615b8ff3c0ca06599ca501b/cpu.cpp#L451 / https://github.com/weidai11/cryptopp/blob/843d74c7c97f9e19a615b8ff3c0ca06599ca501b/cpu.cpp#L393 / https://github.com/weidai11/cryptopp/blob/843d74c7c97f9e19a615b8ff3c0ca06599ca501b/cpu.cpp#L81). This makes building Crypto++ as pure C++ impossible.

@noloader The offending commit is https://github.com/weidai11/cryptopp/commit/e65fa006bec60192112f0a3f23174ef0528f8cc4 which removes the use of the MSVC-specific intrinsics for MSVC versions new enough to actually provide it. I suggest reverting the part of this commit that removes the use of `__cpuidex` and `_xgetbv`.

For OpenMPT, we are building Crypto++ with our own build system in which we define `CRYPTOPP_DISABLE_ASM 1`. This worked with 8.8.0, however it fails now with 8.9.0. Crypto++ is not performance critical in our case, thus we very much prefer a simple C++-only solution in order to keep the build system simple.

We are using Crypto++ to secure our automatic updates on older Windows versions which do not provide the necessary crypto primitives themselves. In the failing case, we use VS2017 to target Windows XP.

**Comment by manxorist** [comment]
In the long term, `CRYPTOPP_DISABLE_ASM` should probably just disable all/most of `cpu.cpp` and all/most of its uses, but that is a way bigger change and kind of a separate issue, IMHO.

**Comment by noloader** [comment]
@manxorist,

I don't quite understand your use case. But looking at the comment that says, "XGETBV64 and CPUID64 are in x64dll.asm," I'm thinking it should be in [`x64masm.asm`](https://github.com/weidai11/cryptopp/blob/843d74c7c97f9e19a615b8ff3c0ca06599ca501b/x64masm.asm) so it is always available to you.

**Comment by manxorist** [comment]
CRYPTOPP_DISABLE_ASM  claims to disable all assembler. x64dll.asm is assembler. Why do I suddenly need to teach my build system how to build asm files with Crypto++ 8.9.0? This was **NOT** the case in 8.8.0. This is a very obvious regression and in conflict with documentation of CRYPTOPP_DISABLE_ASM.

**Comment by noloader** [comment]
> CRYPTOPP_DISABLE_ASM claims to disable all assembler.

That's not quite correct.

`CRYPTOPP_DISABLE_ASM` will disable ASM-based and intrinsic-based algorithm implementations. It has never disabled feature detection. Consider, projects were still getting ASM whether it was MS's `__cpuidex` or Crypto++'s `CPUID64`.

I think moving forward, I'm going to put XGETBV64 and CPUID64 in its own *.asm file to avoid both `x64masm.masm` and `x64dll.masm`. Both files are algorithm implementations, and you should be able to avoid them when `CRYPTOPP_DISABLE_ASM`.

Now, when `CRYPTOPP_DISABLE_ASM` is in effect, maybe we need to provide some stubs for `XGETBV64` and `CPUID64` that just return 0 values.

**Comment by manxorist** [comment]
> > CRYPTOPP_DISABLE_ASM claims to disable all assembler.
> 
> That's not quite correct.
> 
> `CRYPTOPP_DISABLE_ASM` will disable ASM-based and intrinsic-based algorithm implementations. It has never disabled feature detection. Consider, projects were still getting ASM whether it was MS's `__cpuidex` or Crypto++'s `CPUID64`.

But feature detection did not required ASM with MSVC before.

> I think moving forward, I'm going to put XGETBV64 and CPUID64 in its own *.asm file to avoid both `x64masm.masm` and `x64dll.masm`. Both files are algorithm implementations, and you should be able to avoid them when `CRYPTOPP_DISABLE_ASM`.

I fail to see how that will help. The problem is requiring ASM at all, which was introduced by removing the use of the compiler intrinsics.

> Now, when `CRYPTOPP_DISABLE_ASM` is in effect, maybe we need to provide some stubs for `XGETBV64` and `CPUID64` that just return 0 values.

That might also be an option.

Let me try to rephrase the problem I am complaining about here: There is currently no way at all disable the dependency of the C++ code on ASM code for MSVC x64. There was in 8.8.0, which used intrinsics instead of requiring ASM. Intrinsics are less of a problem because they are provided by the compiler itself and do not require build system adoption.

**Comment by noloader** [comment]
How does this look to you:

```diff
$ git diff
diff --git a/cpu.cpp b/cpu.cpp
index 44b57c4d..ff8c5cee 100644
--- a/cpu.cpp
+++ b/cpu.cpp
@@ -388,9 +388,14 @@ extern bool CPU_ProbeSSE2();
 // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85684.
 word64 XGetBV(word32 num)
 {
+// Explicitly handle CRYPTOPP_DISABLE_ASM case.^M
+// https://github.com/weidai11/cryptopp/issues/1240^M
+#if defined(CRYPTOPP_DISABLE_ASM)^M
+       return 0;^M
+^M
 // Required by Visual Studio 2008 and below and Clang on Windows.
 // Use it for all MSVC-compatible compilers.
-#if defined(_M_X64) && defined(CRYPTOPP_MS_STYLE_INLINE_ASSEMBLY)
+#elif defined(_M_X64) && defined(CRYPTOPP_MS_STYLE_INLINE_ASSEMBLY)^M
 
        return XGETBV64(num);
 
@@ -446,9 +451,15 @@ word64 XGetBV(word32 num)
 // cpu.cpp (131): E2211 Inline assembly not allowed in inline and template functions
 bool CpuId(word32 func, word32 subfunc, word32 output[4])
 {
+// Explicitly handle CRYPTOPP_DISABLE_ASM case.^M
+// https://github.com/weidai11/cryptopp/issues/1240^M
+#if defined(CRYPTOPP_DISABLE_ASM)^M
+       output[0] = output[1] = output[2] = output[3] = 0;^M
+       return false;^M
+^M
 // Required by Visual Studio 2008 and below and Clang on Windows.
 // Use it for all MSVC-compatible compilers.
-#if defined(_M_X64) && defined(CRYPTOPP_MS_STYLE_INLINE_ASSEMBLY)
+#elif defined(_M_X64) && defined(CRYPTOPP_MS_STYLE_INLINE_ASSEMBLY)^M
 
        CPUID64(func, subfunc, output);
        return true;
```

**Comment by noloader** [comment]
This should be cleared at [Commit 121014baf0e8](https://github.com/weidai11/cryptopp/commit/121014baf0e8).

**Comment by manxorist** [comment]
Yeah, that certainly works for me.

I still wonder why you prefer to keep only the ASM CPUID implementation even if the compiler supports intrinsics which do not require the use of an external assembler?

It does not look like Crypto++ currently distinguish at a high level between using ASM or intrinsic implementations for its algorithms, but if it someday does, having the intrinsic CPUID implementation has clear advantages (i.e. not requiring tools other than the compiler itself). At a lower level there are more fine-grained macros (i.e. `CRYPTOPP_SSE2_INTRIN_AVAILABLE` vs `CRYPTOPP_SSE2_ASM_AVAILABLE`). Longer term, it might make sense to split some `CRYPTOPP_DISABLE_INTRINSICS` from `CRYPTOPP_DISABLE_ASM`, maybe?

**Comment by noloader** [comment]
> I still wonder why you prefer to keep only the ASM CPUID implementation even if the compiler supports intrinsics which do not require the use of an external assembler?

A single implementation is advantageous for maintenance. We no longer need to differentiate between versions of Visual Studio that added support for something.

**Comment by manxorist** [comment]
> > I still wonder why you prefer to keep only the ASM CPUID implementation even if the compiler supports intrinsics which do not require the use of an external assembler?
> 
> A single implementation is advantageous for maintenance. We no longer need to differentiate between versions of Visual Studio that added support for something.

I somewhat doubt that adding a forced dependency on another build tool was advantageous for maintenance or for your users. It made Crypto++ impossible to build with solely a C++ compiler.

**Comment by noloader** [comment]
> I somewhat doubt that adding a forced dependency on another build tool was advantageous for maintenance or for your users. I

I told you why we did it. You can take it or leave it.

You happened to get lucky in the past with your practices. That broke at Crypto++ 8.9. That's what happens when you wander off the ranch, and start making up your own rules like not compiling or assembling certain source files.

You're the cause of this problem, not the library.

**Comment by manxorist** [comment]
> You're the cause of this problem, not the library.

Your ***C++*** library suddenly ***REQUIRING*** not only a C++ compiler but also an assembler is the root cause of the problem. `CRYPTOPP_DISABLE_ASM` explicitly states `The library will be compiled using C++ only.` (see https://github.com/weidai11/cryptopp/blob/843d74c7c97f9e19a615b8ff3c0ca06599ca501b/config_asm.h#L30C58-L31C29)

**Comment by noloader** [comment]
> Your C++ library suddenly REQUIRING not only a C++ compiler but also an assembler is the root cause of the problem. 

You have always needed a complete toolchain. We have never advertised you did not need one.

You should probably switch to Botan or OpenSSL since you are so dissatisfied at the job we are doing.

**Comment by manxorist** [comment]
> The library will be compiled using C++ only.

yeah ... so?! What am I misunderstanding here?!

**Comment by abdes** [comment]
It looks like we're getting a fix for this... Just chill for a bit while @noloader is preparing it.
Please be nice to OpenSource project maintainers. You cannot imaging how much time we spend dealing with these special cases that seem benign but use enormous amounts of time from us testing and making sure nothing else is broken.
#peace

**Comment by manxorist** [comment]
> It looks like we're getting a fix for this... 

... and I am already using it (see above / <https://github.com/OpenMPT/openmpt/commit/979aece26ac208a7c3a7ff847c3410d70d9cafd1>) ...

> You cannot imaging how much time we spend dealing with these special cases that seem benign but use enormous amounts of time from us testing and making sure nothing else is broken. #peace

As an Open Source maintainer myself I certainly know that.

I still do not get why removing essential features (like not requiring additional tools over solely a C++ compiler) that your users have been relying upon for multiple years is something I should just swallow without complaining about. Pretty please revert <https://github.com/weidai11/cryptopp/commit/49fef81923d9e9ed6706c7ad3a503bfc2aeceedc>.

**Comment by abdes** [comment]
> I still do not get why removing essential features (like not requiring additional tools over solely a C++ compiler) that your users have been relying upon for multiple years is something I should just swallow without complaining about. Pretty please revert [49fef81](https://github.com/weidai11/cryptopp/commit/49fef81923d9e9ed6706c7ad3a503bfc2aeceedc) - this decision IMHO needs far more discussion and approval from other contributors and users.

Just making sure we're on the same page here...
With https://github.com/weidai11/cryptopp/commit/121014baf0e8 you do not need to compile the ASM files anymore. At least, this is what I am doing in my build. Could you confirm that @manxorist ?

**Comment by manxorist** [comment]
> > I still do not get why removing essential features (like not requiring additional tools over solely a C++ compiler) that your users have been relying upon for multiple years is something I should just swallow without complaining about. Pretty please revert [49fef81](https://github.com/weidai11/cryptopp/commit/49fef81923d9e9ed6706c7ad3a503bfc2aeceedc) - this decision IMHO needs far more discussion and approval from other contributors and users.
> 
> Just making sure we're on the same page here... With [121014baf0e8](https://github.com/weidai11/cryptopp/commit/121014baf0e8) you do not need to compile the ASM files anymore. At least, this is what I am doing in my build. Could you confirm that @manxorist ?

Yes, confirmed.

I still take issue with the removed comment for `CRYPTOPP_DISABLE_ASM`.

**Comment by abdes** [comment]
> I still take issue with the removed comment for `CRYPTOPP_DISABLE_ASM`.

My understanding is that the revised comment reflects the actual state of the implementation. How would you rephrase it?

**Comment by manxorist** [comment]
> > I still take issue with the removed comment for `CRYPTOPP_DISABLE_ASM`.
> 
> My understanding is that the revised comment reflects the actual state of the implementation. How would you rephrase it?

I would very much prefer if the implementation would stick to its historic promises, as it did in earlier versions.

If Crypto++ really wants to start requiring building with an assembler, it fails to meet my requirements for a "C++ library".

**Comment by abdes** [comment]
### Known Issues

- [#101](https://github.com/abdes/cryptopp-cmake/issues/101)
  When defining `CRYPTOPP_DISABLE_ASM` to a truthful value, the build will fail if
  the compiler is MSVC on Windows or clang on MSys. A fix has been done on the
  master branch of crypto++
  (https://github.com/weidai11/cryptopp/commit/121014baf0e8).
  cryptopp-cmake tracks the released versions of crypto++ and will be re-released
  when crypto++ is.

### Features

- update source files for crypto++ 8.9.0 ([fa72356](http://github.com/abdes/asap/commit/fa72356c0b89b7346292f2e914ee45bbc742d503)), closes [#88](http://github.com/abdes/asap/issues/88)

### Bug Fixes

- link to omp if requested ([#92](http://github.com/abdes/asap/issues/92)) ([899542f](http://github.com/abdes/asap/commit/899542f1f7f0e55cfa1f41cfb3e9aa9dc157eace))
- make MacOS $(ARCHS_STANDARD) a valid value for CMAKE_OSX_ARCHITECTURES ([6bb6463](http://github.com/abdes/asap/commit/6bb64630c563937e4263b721755c724c0e1ba6e1)), closes [#99](http://github.com/abdes/asap/issues/99)
